### DISCUSSION TITLE
Milvus backup and restore

### DISCUSSION THREAD

**Discussion by sahusasank** [milvus-io/milvus#29630] [Q&A and General discussion]
Hello,

In my env, milvus is running and i am trying to take backup and restore.
I am able to take backup successfully by using "milvus-backup create -n <name>" and able to restore it by "milvus-backup restore -n <name>" but after restoration when i logged in and checking, i am able to see exact number of collections while status is showing "unloaded" and "Approximate entity count" is showing 0. Please help me to fix this.

**Comment by yhmo** [comment]
milvus-backup tool doesn't copy the index for the new collection. You need to create an index for the new collection and load it.

**Comment by sahusasank** [comment]
Thanks for this answer. However i was trying for one collection, attaching the screen shot for collection details and tried to re-index

then it is searching for below:

###
   print(fmt.format("Start Creating index IVF_FLAT"))
    index = {
        "index_type": "IVF_FLAT",
        "metric_type": "L2",
        "params": {"nlist": 128},
    }

    recover_collection.create_index("embeddings", index)

####

in my case, there is no embeddings field then how should i do re-index?

[collection.docx](https://github.com/milvus-io/milvus/files/13816241/collection.docx)

**Reply by yhmo** [reply]
The backup tool does these things:
1. when you call "milvus-backup create", the tool read the collection schema from milvus, and copy its data files to another path
2. when you call "milvus-backup restore:, the tool creates a new collection with the same schema to the old collection, and imports the data files into the new collection

So, the new collection and the old collection have the same schema. If you see different schema, something must b wrong.

**Comment by sahusasank** [comment]
As i said, in one of the milvus, i am taking milvus-backup create and storing backup in Minio

In a new instance of Milvus, when i tried to restore it is restoring all collections while index is showing 0.

As you said i have to re-index but in my above email, you can see re-indexing searching for different field which is not there in both in old as well as new schema. Then how should i do re-index?

**Reply by yhmo** [reply]
If the vector field name is "vector", then` recover_collection.create_index("vector", index)`
Else if the vector field name is "embeddings", then `recover_collection.create_index("embeddings", index)`

**Comment by sahusasank** [comment]
Again i am attaching the difference collection snap in attached doc.
[Uploading collection-diff.docx…]()

**Comment by radhamde** [comment]
Hi Yhmo
Thanks for your response. Please see my approach of backup and restore, and next to my problem - 

1. I have a helm chart based milvus installation, e.g., milvus-1, which contains milvus and minio (e.g., minio-1)
2. I am using milvus-backup tool to take backup (with name my-backup) of my existing milvus collections from milvus-1 installation (after adding credentials of milvus-1 and minio-1, inside backup.yaml file for milvus-backup utility)
3. Next I changing the backup.yaml milvus credential to another helm chart based milvus installation, e.g., milvus-2 . while for minio I am still keeping minio-1 credentials, since I believe the backup is inside minio-1 as we took backup of milvus-1
4. running milvus-backup restore command using my-backup backup copy to restore this backed up collections and their data into milvus-2

On doing the above I am getting all backed up collections created inside milvus-2, while data count are zero for all collections. But if we run the restore with same backup over milvus-1 (where from backup taken), it brings back collections and data as well.

At least this is our experience with example data_prepare.py, which creates hello_milvus and hello_milvus2 collections in milvus-1. Also, data_verify.py is working well to reindex those recovered collections and changing the status of restored collections from Unloaded to Loaded.

Could you please let us know whether there is any issue to restore the backed up collections into a new milvus instance apart from where from backup taken ? 

Also, I would like to know action point in a situation where our original milvus and minio instances are completely destroyed for some reason, where from we took backup and stored in respective minio for same milvus installations, how should we be able to preserve the backed up collections and their data for future restoration to a new milvus instance.

Would be great if you kindly help us solve this matter ASAP.

**Reply by yhmo** [reply]
Some background:
- Milvus stores system metadata in Etcd, and stores data files in S3
- system metadata: collections schema, RBAC user/password, the S3 path of each collection/partition/segment, etc.
- data files: column-based files to stores data of fields for each collection, index files to store index after you call create_index(), bloom-filter files to quickly find entity by id.
- The metada in Etcd must be strongly consist with the data files in S3, otherwise milvus will run into errors.
- Each milvus cluster can only access one minio service by configuration the address in the milvus.yaml
- Each milvus cluster can only access one etcd service by configuration the address in the milvus.yaml


Let's say we have two milvus clusters: milvus-1 and milvus-2
- milvus-1 is using etcd-1 and minio-1
- milvus-2 is using etcd-2 and minio-2


In fact, the milvus-backup tool uses the [bulkinsert](https://milvus.io/docs/bulk_insert.md) interface to import data files into milvus server. The bulkinsert interface requires the data files are accessible by the milvus server, that is to say the data files must be in the same minio server which the milvus server is using.

So, the steps to backup/restore collection from milvus-1 to milvus-2:
1. in the backup.yaml of milvus-backup, set the minio address to be minio-1, set [backupBucketName] equal to [bucketName] 
2. create a backup by milvus-backup, the data files are copied to another path in the minio-1, the path is [backupBucketName]/[backupRootPath]
3. manually copy the data files from minio-1 to minio-2, to the path [backupBucketName]/[backupRootPath]
4. in the backup.yaml of milvus-backup, set the minio address to be minio-2, set [backupBucketName] equal to [bucketName]
5. restore the backup by milvus-backup, the tool will call bulkinsert interface to the milvus-2 and import the data files from minio-2

The key step is the 3, make sure the data files are accessible by milvus-2. If you backup/restore to the same milvus server, then no need to do step 3.
Actually, if the bulkinsert task failed, you will see errors from the log of the milvus-backup tool.

**Reply by yhmo** [reply]
Etcd and minio must be always consistent. 

In the milvus.yaml, the etcd.endpoints and etcd.rootPath determine where the metada is stored, the minio.address and minio.rootPath determine where the data files are stored.


milvus-1 is using etcd-1/rootPath-1 and minio-1/rootPath-1
milvus-2 is using etcd-2/rootPath-2 and minio-2/rootPath-2

The following actions **will lead to disaster**(unexpected errors, data lost, panic):
- restart milvus-x by using etcd-1/rootPath-1 and minio-2/rootPath-2
- or restart milvus-x by using etcd-2/rootPath-2 and minio-1/rootPath-1

**Comment by debjotimallick** [comment]
Hi Yhmo
I am using REST API based backup & restore for Milvus and trying to download the backed up data using the following command (as mentioned in the doc):
curl --location --request GET 'http://localhost:8080/api/v1/get_backup?backup_id <test_backup_id>&backup_name=my_backup'  --header 'Content-Type: application/json'
However, instead of getting the data in return, I am getting the metadata for the backup. 
<img width="943" alt="Screenshot 2024-01-04 184303" src="https://github.com/milvus-io/milvus/assets/45062395/b9644a87-41fe-4a19-8c84-dc94d1e42e74">
Is there any possibility of downloading the data using REST? Also, if at some point minio loses the data it had, how can I feed my downloaded data back into minio for restoration?

**Reply by yhmo** [reply]
"get_backup" command is to get the information of a backup task.
"msg": "success"  ------------------------ this means the backup task is successfully done.
"name": 'my_backup2" ----------------- the backup task name is "my_backup2"
"collection_name": "hello_milvus2" ---- the backuped collection name is "hello_milvus2"

By default, the data files are backuped to S3/minio, under the path "backup", if you browse your s3/minio path, you will see:
 
![Screenshot from 2024-01-05 12-15-54](https://github.com/milvus-io/milvus/assets/2282099/cd9081c2-7814-4fab-945a-921ad5f73022)

If you want to download the data files, use s3/minio client to download.

**Comment by sahusasank** [comment]
Hello, is it possible to take backup in a centralize COS (object Storage) instead connecting to minio-1 then copy to minio-2.
Suppose i have milvus-1,minio-1,etcd-1 is running and milvus-2,minio-2,etcd-2 is running.
I want to take backup from milvus-1 and copy the backup to external Object storage like S3 or IBM cloud Object Storage then restore the backup from external storage to milvus-2?

If yes- can you please provide me the exact steps/procedures and configuration to implement this pls?

**Reply by yhmo** [reply]
Assume the milvus-1 and milvus-2 are using the same s3 address, different buckets.
milvus-1 is using [s3-address]/bucket_1
milvus-2 is using [s3-address]/bucket_2

In the [configuration file](https://github.com/zilliztech/milvus-backup/blob/59d9f5ac8ae99cdf6233bbf2747405054559c82c/configs/backup.yaml#L23) of the milvus-backup tool, we can see the following items:

```
milvus:
  address: 
  port: 

minio:
  address: 
  port:  
  accessKeyID:
  secretAccessKey:

  bucketName: 
  rootPath: 
  
  backupBucketName: 
  backupRootPath: 
```

- To backup data files from milvus-1, we set the bucketName to be bucket_1 and set milvus.address to milvus-1
We want to backup the files to bucket_2 so that the milvus-2 can access the files, so we set the backupBucketName to be bucket_2

```
milvus:
  address: milvus-1
  port: 19530

minio:
  address: 
  port: 9000 
  accessKeyID:
  secretAccessKey:

  bucketName: "bucket_1"
  rootPath: "files"
  
  backupBucketName: "bucket_2"
  backupRootPath: "backup"
```

- Then use /create command to create a backup task
After the files are backuped successfully, you will see they are under the path bucket_2/backup
```
curl --location --request POST 'http://localhost:8080/api/v1/create' \
--header 'Content-Type: application/json' \
--data-raw '{
  "async": true,
  "backup_name": "test_backup",
  "collection_names": [
    "test_collection1"
  ]
}'
```


- Now we want to restore the files into milvus-2, so we need to change milvus.address to milvus-2
Here we change the bucketName to bucket_2 just to mean bucket_2 is the storage of milvus-2. This item is not used during restore
```
milvus:
  address: milvus-2
  port: 19530

minio:
  address: 
  port: 9000 
  accessKeyID:
  secretAccessKey:

  bucketName: "bucket_2"
  rootPath: "files"
  
  backupBucketName: "bucket_2"
  backupRootPath: "backup"
```

- Then use /restore command to restore
Since the backupBucketName is bucket_2, which is the bucket of the milvus-2, the milvus-2 can access this bucket and import the files.
```
curl --location --request POST 'http://localhost:8080/api/v1/restore' \
--header 'Content-Type: application/json' \
--data-raw '{
    "async": true,
    "collection_names": [
    "test_collection1"
  ],
    "collection_suffix": "_bak",
    "backup_name":"test_backup"
}'
```

**Reply by sahusasank** [reply]
ok thanks for this information. 
my milvus is running in SSL mode. can i use the milvus-backup tool to take the backup?

If not, is there any alternative way is thr to take the SSL enabled milvus backup?

**Reply by yhmo** [reply]
The milvus-backup tool supports SSL with milvus server, but it doesn't support TLS with milvus server.
To configure SSL, just modify the backup.yaml, set the user/password here:
https://github.com/zilliztech/milvus-backup/blob/59d9f5ac8ae99cdf6233bbf2747405054559c82c/configs/backup.yaml#L15C17-L15C17

```
milvus:
  authorizationEnabled: true

  user: "root"
  password: "Milvus"
```
